src/main/java/org/schemata/parser/SchemaParser.java

```diff
@@ -1,32 +1,90 @@
-    List<Schema> schemaList = new ArrayList<>();
-    for (GeneratedMessageV3 message : messages) {
-      Descriptors.Descriptor descriptorType = message.getDescriptorForType();
-      String schemaName = descriptorType.getFullName();
-      // Extract all the metadata for the fieldList
-      var fieldList = extractFields(descriptorType.getFields(), schemaName);
-      var schema = this.extractSchema(descriptorType, descriptorType.getFullName(), fieldList);
-      schemaList.add(schema);
+    var descriptors = messages
+            .stream()
+            .map(GeneratedMessageV3::getDescriptorForType)
+            .toList();
+
+    return parseSchemaFromDescriptors(descriptors);
+  }
+
+  public List<Schema> parseSchema(DescriptorProtos.FileDescriptorSet descriptorSet) throws Descriptors.DescriptorValidationException {
+    // build new DAG of file dependencies
+    var dependencyFilenames = new DirectedAcyclicGraph<String, String>(
+            SupplierUtil.createSupplier(String.class),
+            SupplierUtil.createSupplier(String.class), false);
+
+    // populate the graph
+    for (var fileDescriptorProto : descriptorSet.getFileList()) {
+      dependencyFilenames.addVertex(fileDescriptorProto.getName());
+      for (String dependency : fileDescriptorProto.getDependencyList()) {// adding a vertex is idempotent
+        dependencyFilenames.addVertex(dependency);
+        dependencyFilenames.addEdge(dependency, fileDescriptorProto.getName());
+      }
     }
-    return schemaList;
+
+    // key file protos by name for easier lookup when building descriptor objects
+    var fileDescriptorProtosByName = descriptorSet
+            .getFileList()
+            .stream()
+            .collect(Collectors.toMap(DescriptorProtos.FileDescriptorProto::getName, file -> file));
+    var descriptors = new HashMap<String, Descriptors.FileDescriptor>();
+
+    // forEach for DAG is executed in topological sort order
+    for (String filename : dependencyFilenames) {
+      var file = fileDescriptorProtosByName.get(filename);
+      var dependenciesForFile = file
+              .getDependencyList()
+              .stream()
+              .map(descriptors::get).toArray(size -> new Descriptors.FileDescriptor[size]);
+
+      var descriptor = Descriptors.FileDescriptor.buildFrom(file, dependenciesForFile);
+      descriptors.put(filename, descriptor);
+    }
+    
+    var messageDescriptors = descriptors
+            .values()
+            .stream()
+            .flatMap(fileDescriptor -> fileDescriptor.getMessageTypes().stream());
+
+    return parseSchemaFromDescriptors(messageDescriptors.toList());
+  }
+
+  public Schema parseSingleSchema(Descriptors.Descriptor descriptor) {
+    String schemaName = descriptor.getFullName();
+    // Extract all the metadata for the fieldList
+    var fieldList = extractFields(descriptor.getFields(), schemaName);
+    var schema = this.extractSchema(descriptor, descriptor.getFullName(), fieldList);
```

Comment:
possibly we can remove the `this` reference for code consistency.

Reply:
good spotting, fixed in e755485